### PR TITLE
[IDP-622] Allow inherited documents to benefit from indexes provided by their base document

### PR DESCRIPTION
**Pull request by asimmon**
... without throwing an exception

### CODE REVIEW

File: src/.editorconfig [hunk 1]
```diff
@@ -166,7 +166,7 @@ dotnet_diagnostic.CA2002.severity = warning
 dotnet_diagnostic.CA2003.severity = none
 dotnet_diagnostic.CA2004.severity = none
 dotnet_diagnostic.CA2006.severity = none
-dotnet_diagnostic.CA2007.severity = error
+dotnet_diagnostic.CA2007.severity = warning
```
Comment (asimmon) [inline src/.editorconfig:169]: This should have been a warning from the very beginning, it's less annoying during development. Also during CI we already consider warnings as errors

File: src/Workleap.Extensions.Mongo.Tests/MongoIndexerTests.Inheritance.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+using MongoDB.Bson.Serialization.Attributes;
+using MongoDB.Driver;
+using Workleap.Extensions.Mongo.Indexing;
+
+namespace Workleap.Extensions.Mongo.Tests;
+
+public partial class MongoIndexerTests
+{
+    [Fact]
+    public async Task UpdateIndexesAsync_Ignores_Automatically_Inherited_Index_Provider()
```
Comment (asimmon) [inline src/Workleap.Extensions.Mongo.Tests/MongoIndexerTests.Inheritance.cs:11]: The logs displayed in the test also show that the index was created successfully (only once, nothing else):

image

File: src/Workleap.Extensions.Mongo.Tests/MongoIndexerTests.Inheritance.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        await this.Services.GetRequiredService<IMongoIndexer>().UpdateIndexesAsync(documentTypes);
+
+        await AssertIndexes<BaseTestDocumentA>();
+        await AssertIndexes<ChildTestDocumentB>();
+        await AssertIndexes<ChildTestDocumentC>();
+    }
+
+    [BsonKnownTypes(typeof(ChildTestDocumentB), typeof(ChildTestDocumentC))]
+    [MongoCollection("inheritedIndexProviders", IndexProviderType = typeof(BaseTestDocumentAIndexProvider))]
+    private class BaseTestDocumentA : MongoDocument
```
Comment (asimmon) [inline src/Workleap.Extensions.Mongo.Tests/MongoIndexerTests.Inheritance.cs:34]: This was our user's setup.
- A base document class that provides an index provider type, and has two derived types marked as `BsonKnownTypes`
- The two derived types don't have to be decorated with `MongoCollection` nor have to provide an index provider.

The derived types here benefit from the indexes declared by their base type.

File: src/Workleap.Extensions.Mongo.Tests/MongoIndexerTests.cs [hunk 4]
```diff
@@ -5,7 +5,7 @@
 
 namespace Workleap.Extensions.Mongo.Tests;
 
-public sealed class MongoIndexerTests : BaseIntegrationTest<MongoFixture>
+public partial class MongoIndexerTests : BaseIntegrationTest<MongoFixture>
```
Comment (asimmon) [inline src/Workleap.Extensions.Mongo.Tests/MongoIndexerTests.cs:8]: I believe it increased readability to split this class into partial classes. The whole class is about indexing.

File: src/Workleap.Extensions.Mongo/Indexing/IndexRegistry.cs [hunk 5]
```diff
@@ -6,8 +6,14 @@ namespace Workleap.Extensions.Mongo.Indexing;
 /// <summary>
 /// Associates a concrete <see cref="IMongoDocument"/> class with its single <see cref="MongoIndexProvider"/>.
 /// </summary>
-internal sealed class IndexRegistry : Dictionary<Type, Type>
+internal sealed class IndexRegistry : Dictionary<string, DocumentTypeEntry>
```
Comment (asimmon) [inline src/Workleap.Extensions.Mongo/Indexing/IndexRegistry.cs:9]: When designing the library two years ago, I made a mistake. I believed that developers would only declare ONE single C# class for exactly ONE MongoDB collection.

Time has shown that developers can create as many C# document types that could be mapped to a same collection, as long as they have a similar shape. Most of the time, they only contain a subset of the properties.

I never thought that someone would use inheritance, combined with the index provider, to target a same collection.

So this `IndexRegistry` was fundamentally wrong by assuming that its dictionary key was the C# document type. It's supposed to be the collection name. Only the collection is unique. Also here now, we make sure that for one collection, there's only one index provider type as well.

**Review by Zav — APPROVED** [review]

**Pull request merged**